File: practica1/backend1/src/routes/purchase.js
import express from "express";
import { pool } from "../db.js";

const router = express.Router();

/**
 * POST /purchase
 * body: { buyerId, artworkId }
 *
 * Llama al SP transaccional sp_purchase(buyerId, artworkId)
 * La lógica de negocio (locks, saldo, transferencia, notificaciones)
 * vive en la base de datos.
 */
router.post("/", async (req, res) => {
  try {
    const { buyerId, artworkId } = req.body || {};
    const buyer = Number(buyerId);
    const artId = Number(artworkId);

    if (!buyer || !artId) {
      return res.status(400).json({ error: "buyerId y artworkId son requeridos" });
    }

    // Ejecuta el SP (él hace todo: valida, mueve saldos, transfiere, notifica)
    await pool.query("CALL sp_purchase(?, ?)", [buyer, artId]);

    // Si llegó aquí, la compra se concretó
    return res.json({ ok: true, artworkId: artId, buyerId: buyer });
  } catch (e) {
    // Errores de negocio señalados con SIGNAL 45000 desde el SP
    if (e && (e.errno === 1644 || e.sqlState === "45000")) {
      // Mensajes posibles desde el SP/Triggers:
      // - 'La obra no existe'
      // - 'La obra no está disponible'
      // - 'No puedes comprar tu propia obra'
      // - 'Saldo insuficiente'
      // - 'Vendedor no existe' / 'Comprador no existe'
      return res.status(409).json({ error: e.sqlMessage || "Operación inválida" });
    }

    // Deadlocks o timeouts (opcionales)
    if (e && e.code === "ER_LOCK_DEADLOCK") {
      return res.status(409).json({ error: "Conflicto de concurrencia, intenta de nuevo" });
    }

    console.error("POST /purchase error:", e);
    return res.status(500).json({ error: "No se pudo completar la compra" });
  }
});

export default router;

File: practica1/backend2/src/routes/purchase.py
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel
from ..db import db

router = APIRouter()

class PurchaseRequest(BaseModel):
    buyerId: int
    artworkId: int

@router.post("/")
async def purchase_artwork(request: PurchaseRequest):
    """
    Procesar compra de obra de arte
    
    Llama al SP transaccional sp_purchase(buyerId, artworkId)
    La lógica de negocio (locks, saldo, transferencia, notificaciones)
    vive en la base de datos.
    """
    try:
        buyer = request.buyerId
        art_id = request.artworkId

        if not buyer or not art_id:
            raise HTTPException(
                status_code=400, 
                detail="buyerId y artworkId son requeridos"
            )

        # Ejecuta el SP (él hace todo: valida, mueve saldos, transfiere, notifica)
        await db.execute_procedure('sp_purchase', [buyer, art_id])

        # Si llegó aquí, la compra se concretó
        return {
            "ok": True,
            "artworkId": art_id,
            "buyerId": buyer
        }

    except Exception as e:
        error_msg = str(e)
        
        # Errores de negocio señalados con SIGNAL 45000 desde el SP
        if ('45000' in error_msg or 
            'La obra no existe' in error_msg or
            'La obra no está disponible' in error_msg or
            'No puedes comprar tu propia obra' in error_msg or
            'Saldo insuficiente' in error_msg or
            'Vendedor no existe' in error_msg or
            'Comprador no existe' in error_msg):
            
            raise HTTPException(
                status_code=409, 
                detail=error_msg if 'La obra' in error_msg or 'No puedes' in error_msg or 'Saldo' in error_msg 
                       else "Operación inválida"
            )

        # Deadlocks o timeouts
        if 'Deadlock' in error_msg or 'Lock wait timeout' in error_msg:
            raise HTTPException(
                status_code=409, 
                detail="Conflicto de concurrencia, intenta de nuevo"
            )

        print(f"POST /purchase error: {e}")
        raise HTTPException(
            status_code=500, 
            detail="No se pudo completar la compra"
        )
